bookkeeper-server/src/main/java/org/apache/bookkeeper/meta/CleanupLedgerManager.java

```diff
@@ -75,11 +76,17 @@ public LedgerRange next() throws IOException {
         }
     }
 
+    private static final CompletableFuture<Versioned<LedgerMetadata>> closedPromise = new CompletableFuture<>();
```

Comment:
don't do this.  you can have a static method creating the future, but never keep it as a static variable.
every future should be a brand new instance, otherwise, all the callbacks will be registered here.

Reply:
https://github.com/bpupadhyaya/openjdk-8/blob/45af329463a45955ea2759b89cb0ebfe40570c3f/jdk/src/share/classes/java/util/concurrent/CompletableFuture.java#L2015

If the future is complete, nothing is queued. It's executed straight away. I've no problem creating a future each time, this isn't fast path code, but it is safe to use a future like this.

Reply:
Yeah. I would prefer creating a future each time, that would be a clean path, and will be no surprise from gc perspective.